src/libcore/pin.rs

```diff
@@ -184,22 +348,74 @@ impl<P: Deref> Pin<P> {
     /// Construct a new `Pin` around a reference to some data of a type that
     /// may or may not implement `Unpin`.
     ///
+    /// If `pointer` dereferences to an `Unpin` type, `Pin::new` should be used
+    /// instead.
+    ///
     /// # Safety
     ///
     /// This constructor is unsafe because we cannot guarantee that the data
-    /// pointed to by `pointer` is pinned. If the constructed `Pin<P>` does
+    /// pointed to by `pointer` is pinned, meaning that the data will not be moved or
+    /// its storage invalidated until it gets dropped. If the constructed `Pin<P>` does
     /// not guarantee that the data `P` points to is pinned, constructing a
-    /// `Pin<P>` is undefined behavior.
+    /// `Pin<P>` is unsafe.
```

Comment:
What do you mean by "unsafe" here and why the change from UB?

Reply:
Well, it's not UB. There's nothing in the Rust abstract machine that says this is wrong.

It's like saying that calling `set_len()` on a `Vec` is UB. It never is. But later operations might be.

Reply:
Right, in that case "...may lead to undefined behavior in later operations." may be clearer. "unsafe" doesn't sound like an API violation has happened.

Reply:
I made it

> If the constructed `Pin<P>` does
> not guarantee that the data `P` points to is pinned, that is a violation of
> the API contract and may lead to undefined behavior in later (safe) operations.